src/BSON/Decimal128.c

```diff
@@ -0,0 +1,205 @@
+  | Copyright 2013-2015 MongoDB, Inc.                                         |
+  |                                                                           |
+  | Licensed under the Apache License, Version 2.0 (the "License");           |
+  | you may not use this file except in compliance with the License.          |
+  | You may obtain a copy of the License at                                   |
+  |                                                                           |
+  | http://www.apache.org/licenses/LICENSE-2.0                                |
+  |                                                                           |
+  | Unless required by applicable law or agreed to in writing, software       |
+  | distributed under the License is distributed on an "AS IS" BASIS,         |
+  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  |
+  | See the License for the specific language governing permissions and       |
+  | limitations under the License.                                            |
+  +---------------------------------------------------------------------------+
+  | Copyright (c) 2014-2015 MongoDB, Inc.                                     |
+  +---------------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+#	include "config.h"
+#endif
+
+/* External libs */
+#include <bson.h>
+#include <mongoc.h>
+
+/* PHP Core stuff */
+#include <php.h>
+#include <php_ini.h>
+#include <ext/standard/info.h>
+#include <Zend/zend_interfaces.h>
+#include <ext/spl/spl_iterators.h>
+/* Our Compatability header */
+#include "phongo_compat.h"
+
+/* Our stuffz */
+#include "php_phongo.h"
+#include "php_bson.h"
+
+
+PHONGO_API zend_class_entry *php_phongo_decimal128_ce;
+
+zend_object_handlers php_phongo_handler_decimal128;
+
+/* {{{ proto BSON\Decimal128 Decimal128::__construct(string $value)
+   Construct a new BSON Decimal128 type */
+PHP_METHOD(Decimal128, __construct)
+{
+	php_phongo_decimal128_t *intern;
+	zend_error_handling      error_handling;
+	char                    *value;
+	phongo_zpp_char_len      value_len;
+
+
+	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling TSRMLS_CC);
+	intern = Z_DECIMAL128_OBJ_P(getThis());
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &value, &value_len) == FAILURE) {
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+	zend_restore_error_handling(&error_handling TSRMLS_CC);
+
+	if ( ! bson_decimal128_from_string(value, &intern->decimal)) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error parsing decimal string: %s", value);
+		return;
+	}
+}
+/* }}} */
+
+/* {{{ proto void Decimal128::__toString()
+    */
+PHP_METHOD(Decimal128, __toString)
+{
+	php_phongo_decimal128_t *intern;
+	char                     outbuf[BSON_DECIMAL128_STRING] = "";
```

Comment:
It's a bit of defensive programming, since `bson_decimal128_to_string()` is void. I wasn't sure if it might leave the output variable untouched on some error. That said, I think it's best to just `memset()` the whole thing.